### PR TITLE
[USM] Avoid https flaky tests

### PR DESCRIPTION
**Pull request by nplanel**
### What does this PR do?

Avoid https flaky tests, by checking in 2 stages (USM, then NPM) 
matching the connection, tags and client pid

### Motivation

https flaky tests

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/tracer/tracer_usm_linux_test.go [hunk 1]
```diff
@@ -285,43 +286,59 @@ func testHTTPSLibrary(t *testing.T, fetchCmd []string, prefetchLibs []string) {
...
+		allConnections = append(allConnections, payload.Conns...)
 		for key, stats := range payload.HTTP {
+			if key.Path.Content != "/200/foobar" {
+				continue
+			}
 			req, exists := stats.Data[200]
 			if !exists {
-				continue
+				errorLogs := fmt.Sprintf("http %# v stats %# v", krpretty.Formatter(key), krpretty.Formatter(stats.Data))
+				t.Logf("%s", errorLogs)
```
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:304]: why do we need both sprintf and logf?

File: pkg/network/tracer/tracer_usm_linux_test.go [hunk 2]
```diff
@@ -285,43 +286,59 @@ func testHTTPSLibrary(t *testing.T, fetchCmd []string, prefetchLibs []string) {
...
+				t.Fail()
```
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:305]: is it reasonable to call it inside require.Eventually?

File: pkg/network/tracer/tracer_usm_linux_test.go [hunk 3]
```diff
@@ -285,43 +286,59 @@ func testHTTPSLibrary(t *testing.T, fetchCmd []string, prefetchLibs []string) {
...
 
 			statsTags := req.StaticTags
 			// debian 10 have curl binary linked with openssl and gnutls but use only openssl during tls query (there no runtime flag available)
 			// this make harder to map lib and tags, one set of tag should match but not both
 			if key.Path.Content == "/200/foobar" && (statsTags == network.ConnTagGnuTLS || statsTags == network.ConnTagOpenSSL) {
 				t.Logf("found tag 0x%x %s", statsTags, network.GetStaticTags(statsTags))
+				httpKey = key
+				return true
+			}
+			if key.Path.Content == "/200/foobar" {
```
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:316]: redundant check, already done in 298

File: pkg/network/tracer/tracer_usm_linux_test.go [hunk 4]
```diff
@@ -285,43 +286,59 @@ func testHTTPSLibrary(t *testing.T, fetchCmd []string, prefetchLibs []string) {
...
+		}
+		return false
+	}, 15*time.Second, 5*time.Second, "couldn't find USM HTTPS stats")
+
+	// socket filter is not supported on fentry tracer, so can't classify TLS protocol
 
-				// socket filter is not supported on fentry tracer
-				if fentryTracerEnabled {
-					// so we return early if the test was successful until now
+	if !fentryTracerEnabled {
```
Comment (guyarb) [inline pkg/network/tracer/tracer_usm_linux_test.go:325]: it is guaranteed that we are not running with fentry enabled
https://github.com/DataDog/datadog-agent/pull/16910

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/tracer/tracer_usm_linux_test.go (hunk 2 above)
Comment (nplanel) [reply pkg/network/tracer/tracer_usm_linux_test.go]: I moved to t.Errorf() and return false, that's doing the same thing

File: pkg/network/tracer/tracer_usm_linux_test.go (hunk 4 above)
Comment (nplanel) [reply pkg/network/tracer/tracer_usm_linux_test.go:322]: shall we backport #16910 to 7.45 (and as well this one) ?

File: pkg/network/tracer/tracer_usm_linux_test.go (hunk 2 above)
Comment (guyarb) [reply pkg/network/tracer/tracer_usm_linux_test.go]: t.Error calls t.Fail

File: pkg/network/tracer/tracer_usm_linux_test.go (hunk 4 above)
Comment (guyarb) [reply pkg/network/tracer/tracer_usm_linux_test.go]: IMHO no, we're merging only broken tests, flaky tests are not urgent

File: pkg/network/ebpf/c/tracer/tracer.h [hunk 5]
```diff
@@ -43,8 +43,6 @@ typedef struct {
     __u64 sent_packets;
     __u64 recv_packets;
     __u8 direction;
-    // keep the conn_tags u8 to keep the struct slim
-    __u8 conn_tags;
```
Comment (brycekahle) [inline pkg/network/ebpf/c/tracer/tracer.h:47]: Why are you removing the tags?

Comment (nplanel) [reply pkg/network/ebpf/c/tracer/tracer.h:47]: we don't use it anymore as Pedro merged the `protocol_stack` and forgot to cleanup this

**Review by hmahmood — APPROVED** [review]

File: pkg/network/ebpf/c/protocols/classification/dispatcher-helpers.h [hunk 6]
```diff
@@ -89,7 +89,9 @@ static __always_inline void protocol_dispatcher_entrypoint(struct __sk_buff *skb
         return;
     }
 
-    // TODO: consider adding early return if `is_layer_known(stack, LAYER_ENCRYPTION)`
+    if (is_fully_classified(stack) || is_protocol_layer_known(stack, LAYER_ENCRYPTION)) {
+        return;
+    }
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/classification/dispatcher-helpers.h:94]: cc @p-lambert please review as well

File: pkg/network/ebpf/c/protocols/http/http.h [hunk 7]
```diff
@@ -164,6 +164,14 @@ static __always_inline bool http_allow_packet(http_transaction_t *http, struct _
         return false;
     }
 
+    protocol_stack_t *stack = get_protocol_stack(&http->tup);
+    if (!stack) {
+        return false;
+    }
+    if (is_fully_classified(stack) || is_protocol_layer_known(stack, LAYER_ENCRYPTION)) {
+        return false;
+    }
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http/http.h:173]: please add a comment about the changes

File: pkg/network/tracer/connection/tracer.go [hunk 8]
```diff
@@ -557,7 +557,6 @@ func populateConnStats(stats *network.ConnectionStats, t *netebpf.ConnTuple, s *
 		Application: protocols.Application(s.Protocol_stack.Application),
 		Encryption:  protocols.Encryption(s.Protocol_stack.Encryption),
 	}
-	stats.StaticTags |= uint64(s.Conn_tags)
```
Comment (guyarb) [inline pkg/network/tracer/connection/tracer.go:560]: why did you remove that?

Comment (nplanel) [reply pkg/network/tracer/connection/tracer.go:563]: https://github.com/DataDog/datadog-agent/pull/16940/files#r1194171920

File: pkg/network/ebpf/c/protocols/http/http.h [hunk 9]
```diff
@@ -164,10 +164,13 @@ static __always_inline bool http_allow_packet(http_transaction_t *http, struct _
...
-    // process it if the packet represents a TCP termination
-    bool empty_payload = skb_info->data_off == skb->len;
-    if (empty_payload || http->tup.sport == HTTPS_PORT || http->tup.dport == HTTPS_PORT) {
+    protocol_stack_t *stack = get_protocol_stack(&http->tup);
+    if (!stack) {
+        return false;
+    }
+    if (is_fully_classified(stack) || is_protocol_layer_known(stack, LAYER_ENCRYPTION)) {
+        // if this is an encrypted packet, we only
+        // process it if the packet represents a TCP termination
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http/http.h:173]: why did you remove the old checks? especially empty payload

File: pkg/network/ebpf/c/protocols/http/http.h (hunk 7 above)
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/http/http.h]: not sure I follow here, but why don't we need to flush the transaction on rst/fin here?

File: pkg/network/ebpf/c/protocols/http/http.h (hunk 9 above)
Comment (nplanel) [reply pkg/network/ebpf/c/protocols/http/http.h:173]: it's more accurate to pass only tcp fin/rst for encrypted flow not based on port
I will re-add the empty payload check here

**Review by guyarb — APPROVED** [review]
